### ISSUE TITLE
Synchronous and Asynchronous mode in Carla_Apollo_Bridge

### ISSUE THREAD

**Issue by AOOOOOA** [guardstrikelab/carla_apollo_bridge#198]
**Note:** If you have a general support question and are looking for a quicker response, please checkout our discord channel for answers from the community:
https://bbs.carla.org.cn/ask

## In what area(s)?

> /area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Ask your question here

Hi,

Thanks for your brilliant work!  However, I have a few questions about the synchronous and asynchronous mode. 

Carla has sync mode and async mode with fixed or varied time step as they introduced:
https://carla.readthedocs.io/en/latest/adv_synchrony_timestep/#:~:text=By%20default%2C%20CARLA%20runs%20in,only%20one%20client%20should%20tick.

The main.py in Carla_Apollo_Bridge adopts the async mode with varied time steps:
![image](https://github.com/guardstrikelab/carla_apollo_bridge/assets/38175747/4be85d63-6357-4db3-bb3a-92b1774685f3)

Is this a mandatory setting to guarantee the normal running of Apollo in Carla? Can I change it to sync mode with a fixed time step? 

Thanks for your kind answer!

**Comment by hlyyy** [comment]
Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?

**Comment by AOOOOOA** [comment]
> Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?

Hi, I took a closer look at the code and found out they have set the mode to sync+fixed time-stamp. The figure above is the final setting after running. What do you mean of the "useless", the vehicle in Carla can not execute the same command as in DreamView?

**Comment by AOOOOOA** [comment]
> > > Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?
> > 
> > 
> > Hi, I took a closer look at the code and found out they have set the mode to sync+fixed time-stamp. The figure above is the final setting after running. What do you mean of the "useless", the vehicle in Carla can not execute the same command as in DreamView?
> 
> Not really. I created a synchronous client based on the bridge to control the movement of the car. But the sync mode didn't work. When I haven't run world.tick(), the car is already moving. I guess it's due to the sync mode in the bridge you mentioned.

The car already moving in both Apollo DreamView and Carla ?

**Comment by AOOOOOA** [comment]
> > > > > Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?
> > > > 
> > > > 
> > > > Hi, I took a closer look at the code and found out they have set the mode to sync+fixed time-stamp. The figure above is the final setting after running. What do you mean of the "useless", the vehicle in Carla can not execute the same command as in DreamView?
> > > 
> > > 
> > > Not really. I created a synchronous client based on the bridge to control the movement of the car. But the sync mode didn't work. When I haven't run world.tick(), the car is already moving. I guess it's due to the sync mode in the bridge you mentioned.
> > 
> > 
> > The car already moving in both Apollo DreamView and Carla ?
> 
> Yes. And it is not controlled by the sync client I created.

It is wired. Could you please provide the code so that I can reproduce the issue?

**Comment by hlyyy** [comment]
> > > > > > Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?
> > > > > 
> > > > > 
> > > > > Hi, I took a closer look at the code and found out they have set the mode to sync+fixed time-stamp. The figure above is the final setting after running. What do you mean of the "useless", the vehicle in Carla can not execute the same command as in DreamView?
> > > > 
> > > > 
> > > > Not really. I created a synchronous client based on the bridge to control the movement of the car. But the sync mode didn't work. When I haven't run world.tick(), the car is already moving. I guess it's due to the sync mode in the bridge you mentioned.
> > > 
> > > 
> > > The car already moving in both Apollo DreamView and Carla ?
> > 
> > 
> > Yes. And it is not controlled by the sync client I created.
> 
> It is wired. Could you please provide the code so that I can reproduce the issue?

I guess it is because Carla only allow one client to use sync mode. And the bridge has already used a sync mode. So the sync client i created didn't work. I was also confused when i encountered this problem.

**Comment by AOOOOOA** [comment]
> > > > > > > Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?
> > > > > > 
> > > > > > 
> > > > > > Hi, I took a closer look at the code and found out they have set the mode to sync+fixed time-stamp. The figure above is the final setting after running. What do you mean of the "useless", the vehicle in Carla can not execute the same command as in DreamView?
> > > > > 
> > > > > 
> > > > > Not really. I created a synchronous client based on the bridge to control the movement of the car. But the sync mode didn't work. When I haven't run world.tick(), the car is already moving. I guess it's due to the sync mode in the bridge you mentioned.
> > > > 
> > > > 
> > > > The car already moving in both Apollo DreamView and Carla ?
> > > 
> > > 
> > > Yes. And it is not controlled by the sync client I created.
> > 
> > 
> > It is wired. Could you please provide the code so that I can reproduce the issue?
> 
> I guess it is because Carla only allow one client to use sync mode. And the bridge has already used a sync mode. So the sync client i created didn't work. I was also confused when i encountered this problem.

No, as it said in Carla Documentation, the sync mode is applied to the global setting. Only one client needs to call the world.tick().

**Comment by hlyyy** [comment]
> > > > > > > > Hi, I have the same question. I found that creating a synchronous client based on the bridge is useless. Do you have any other solution to solve this problem?
> > > > > > > 
> > > > > > > 
> > > > > > > Hi, I took a closer look at the code and found out they have set the mode to sync+fixed time-stamp. The figure above is the final setting after running. What do you mean of the "useless", the vehicle in Carla can not execute the same command as in DreamView?
> > > > > > 
> > > > > > 
> > > > > > Not really. I created a synchronous client based on the bridge to control the movement of the car. But the sync mode didn't work. When I haven't run world.tick(), the car is already moving. I guess it's due to the sync mode in the bridge you mentioned.
> > > > > 
> > > > > 
> > > > > The car already moving in both Apollo DreamView and Carla ?
> > > > 
> > > > 
> > > > Yes. And it is not controlled by the sync client I created.
> > > 
> > > 
> > > It is wired. Could you please provide the code so that I can reproduce the issue?
> > 
> > 
> > I guess it is because Carla only allow one client to use sync mode. And the bridge has already used a sync mode. So the sync client i created didn't work. I was also confused when i encountered this problem.
> 
> No, as it said in Carla Documentation, the sync mode is applied to the global setting. Only one client needs to call the world.tick().

I understand! But how do we control the simulation and the car based on the bridge? I mean getting the vehicle information or controlling other actors movement in each frame.

**Comment by persistisok** [comment]
> Hi,
> 
> Thanks for your brilliant work! However, I have a few questions about the synchronous and asynchronous mode.
> 
> Carla has sync mode and async mode with fixed or varied time step as they introduced: https://carla.readthedocs.io/en/latest/adv_synchrony_timestep/#:~:text=By%20default%2C%20CARLA%20runs%20in,only%20one%20client%20should%20tick.
> 
> The main.py in Carla_Apollo_Bridge adopts the async mode with varied time steps: ![image](https://github.com/guardstrikelab/carla_apollo_bridge/assets/38175747/4be85d63-6357-4db3-bb3a-92b1774685f3)
> 
> Is this a mandatory setting to guarantee the normal running of Apollo in Carla? Can I change it to sync mode with a fixed time step?
> 
> Thanks for your kind answer! [@XiaoFei9704](https://github.com/XiaoFei9704) [@synkrotron](https://github.com/synkrotron) [@meua](https://github.com/meua)

The code after finally sets the bridge to asynchronous mode only after it has finished running. The bridge actually runs in synchronous mode with a fixed step size. The same problem that bothers me is that in addition to the bridge, other clients need to use synchronous mode to load scenes, and using multiple clients synchronously seems to cause some problems.
